### DISCUSSION TITLE
Is it possible to create a container format and then convert other formats to that format?

### DISCUSSION THREAD

**Discussion by camenduru** [huggingface/safetensors#137] [Ideas]
All contain the same thing, but in different boxes 📦. If this is correct, can you please create a single container and converters to convert everything below to that container?

| Format                  | Safe | Zero-copy | Lazy loading | No file size limit | Layout control | Flexibility | Bfloat16
| ----------------------- | --- | --- | --- | --- | --- | --- | --- |
| pickle (PyTorch)        | ✗ | ✗ | ✗ | 🗸 | ✗ | 🗸 | 🗸 |
| H5 (Tensorflow)         | 🗸 | ✗ | 🗸 | 🗸 | ~ | ~ | ✗ |
| SavedModel (Tensorflow) | 🗸 | ✗ | ✗ | 🗸 | 🗸 | ✗ | 🗸 |
| MsgPack (flax)          | 🗸 | 🗸 | ✗ | 🗸 | ✗ | ✗ | 🗸 |
| Protobuf (ONNX)         | 🗸 | ✗ | ✗ | ✗ | ✗ | ✗ | 🗸 |
| Cap'n'Proto             | 🗸 | 🗸 | ~ | 🗸 | 🗸 | ~ | ✗ |
| Arrow                   | ? | ? | ? | ? | ? | ? | ✗ |
| Numpy (npy,npz)         | 🗸 | ? | ? | ✗ | 🗸 | ✗ | ✗ |
| SafeTensors             | 🗸 | 🗸 | 🗸 | 🗸 | 🗸 | ✗ | 🗸 |

**Comment by julien-c** [comment]
for now you can already convert from `pytorch pickle` to `safetensors` from this Space: https://huggingface.co/spaces/safetensors/convert

**Reply by camenduru** [reply]
@julien-c thanks ❤ 
when I can convert `MsgPack (flax) to SafeTensors` also `Protobuf (ONNX) to SafeTensors` when we will unite all the containers to one not just `pickle (PyTorch)`  every new framework coming with a new container type 😕

**Reply by julien-c** [reply]
ONNX embeds the actual graph so there won't be a way to convert to safetensors without other files:

schematically ONNX = safetensors + actual model **code**

**Reply by camenduru** [reply]
thanks ❤ do you have a plan to convert `MsgPack (flax) to SafeTensors`

**Reply by camenduru** [reply]
I added `ckpt to safetensors`  to here https://huggingface.co/spaces/camenduru/converter  🎉 I wish there is a plain [converter](https://github.com/huggingface/safetensors/blob/main/bindings/python/convert.py) instead of adding huggingface_hub functionality inside converter  `pt to safetensors` will take time 😋

**Reply by Narsil** [reply]
@camenduru would that work ? https://gist.github.com/Narsil/d5b0d747e5c8c299eb6d82709e480e3d

I might update `convert.py` but there's more subtlety to it when it comes to converting `transformers` models.

`model.safetensors` files can be loaded by TF, Flax AND pytorch, regardless. And it should be usually faster than all 3 (mostly because msgpack is rather slow for some reason).

BUT and it's a big but, currently by having 3 different files, the actual weights are not necessarily saved exactly in the same shape. The easiest example to understand is that Pytorch uses `NCHW` format (channel first) while TF/Jax use `NHWC` (channel last) format.
So currently for `transformers` conversion you would need to add some `metadata` to specify in which format are the tensors (`pt` / `tf` etc..)

Not something we can't handle, but something to be considerate about and which might not belong into `convert.py` and be better suited elsewhere.

**Reply by Narsil** [reply]
Also as @julien-c said, not everything is directly copiable as `safetensors`.

`msgpack` supports nesting (the script I gave you unpack them, and it should work with `transformers` but maybe someone used `msgpack` to save stuff very different from tensors.
Same for `pickle` we can't convert arbitrary pickles, only ones that contain only Pytorch Tensors (not sparse, not using shared memory).
ONNX and SavedModel contain the call graph of the model.

`Cap'n'proto` is more general than safetensors and can definitely contain much more than tensors.
`numpy` and `npz` should be relatively straightforward to write a converter.

**Reply by camenduru** [reply]
thanks @Narsil  ❤ this https://gist.github.com/Narsil/d5b0d747e5c8c299eb6d82709e480e3d looks super cool 🎉 I will try it with this  https://huggingface.co/camenduru/plushies

**Reply by camenduru** [reply]
hi @Narsil 👋 I converted to `safetensors.flax` but I realized there is no diffusers implementation for `safetensors.flax` how can I convert `safetensors.flax to safetensors.torch`

**Reply by Narsil** [reply]
Could you share a bit more precisely what model and what file you want to convert, and how you intend to use it please ?

Just trying to avoid the issues I mention here:

> BUT and it's a big but, currently by having 3 different files, the actual weights are not necessarily saved exactly in the same shape. The easiest example to understand is that Pytorch uses NCHW format (channel first) while TF/Jax use NHWC (channel last) format.
> So currently for transformers conversion you would need to add some metadata to specify in which format are the tensors (pt / tf etc..)

(Afaik, this is not handled in `diffusers`  yet, as flax support works but is not as active as torch)

**Reply by camenduru** [reply]
https://huggingface.co/camenduru/plushies please convert `safetensors.flax to safetensors.torch` then I can use with `diffusers`

**Reply by camenduru** [reply]
Important question is can safetensors convert safetensors.flax to safetensors.torch?

**Reply by camenduru** [reply]
https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_flax_pytorch_utils.py#L242-L352 maybe like this

**Reply by Narsil** [reply]
Well, this is very specific to `transformers` since operations happen based on weights names, so you would have to know how you named your tensors to know how to convert back&forth.

There are many ways to go about this.

>  safetensors.flax to safetensors.torch

There's no such concept as `safetensors.flax` nor `safetensors.torch`. Safetensors, saves tensors, and doesn't care where they come from or from what lib. The differences exists purely in what tensors (as in preferred shapes) both libraries use. (As I mention, it's a choice done within the libraries which revolve around channel order and sometimes transposed tensors, and is all linked on how the libraries made some choices). It has nothing to do with tensors themselves.

So you can load any safetensors file in Flax or PyTorch without any issues. But you might encounter issues when you try to associate said tensors to the weights of your NN. 
for instance Convolution weights will probably require some transpose if they are loaded from flax and used onto PyTorch.

This will always have a cost in terms of time (since you basically have to duplicate the tensor), but since safetensors seems to load much faster than Flax, it should still be a net win.

**Reply by camenduru** [reply]
thanks @Narsil ❤ I thought safetensors can do cross conversion 😕 I was wrong.  we are trying to solve [this](https://github.com/huggingface/diffusers/pull/1241) problem for more than a month 🥺 can you solve it

**Reply by Narsil** [reply]
Isn't this what you were  looking for ? https://github.com/huggingface/diffusers/pull/1739

**Reply by camenduru** [reply]
no, check the tittle "Allow Flax model to save as pytorch (`If converted from pytorch`)"

**Reply by camenduru** [reply]
![download](https://user-images.githubusercontent.com/54370274/209425972-83675ee7-4a0b-459b-b5c6-8c337633fcd4.png)
https://github.com/camenduru/diffusers/blob/from_flax_v2/src/diffusers/modeling_pytorch_flax_utils.py
